Update to `grants` documentation (#1707)
dbt-labs/docs.getdbt.com: website/docs/reference/resource-configs/grants.md

 
For more information on hooks, see [Hooks & operations](/building-a-dbt-project/hooks-operations).

## Definition

You can use the `grants` field to set permissions or grants for a resource. When you run a model, seed or seed, or snapshot a snapshot, dbt will run `grant` and/or `revoke` statements to ensure that the permissions on the database object match the `grants` you have configured on the resource.

> Like all configurations, `grants` will be included in dbt project metadata, including [the manifest artifact](dbt-artifacts/manifest-json).

Reviewer: will these be surfaced in dbt docs page for a given model now as well?

Author: Not yet - but they could be! I think it would take just a bit of `dbt-docs` code to make that happen.

</TabItem>
</Tabs>

### Grant config inheritance

> If `grants` are set for the same model in multiple places, such as in `dbt_project.yml` and in a more-specific `.sql` or `.yml` file, the default behavior is to "merge and clobber" each privilege. That is, dbt will replace the less-specific set of grantees with the more-specific set of grantees.

Reviewer: this is so cool! utopian idea -- perhaps there should be a warning at least when there's overlap b/w model and folder grants? imagine a message like:
> You've granted `user_c` `READ` rights to the entire `mart_finance/` directory so it is redundant to grant them access specifically to the following models:
> - `fin.F22_forecast_summary`
> Would you like me to update your dbt project to represent that simplification?

idk I can imagine a world in which there's enough redundant grants to the extent that it obfuscates what's going on? 🤷🏻

Author: Hmm I could imagine something like that! Or maybe even: a way to recommend new folder-level configs to DRY up per-model configs. (Or to recommend folder-level reorganization...)

There's some more ideation about how configs ought to be merged (updated/clobbered/added/removed) when defined in multiple places: https://github.com/dbt-labs/dbt-core/issues/4108

> If `grants` are set for the same model in multiple places, such as in `dbt_project.yml` and in a more-specific `.sql` or `.yml` file, the default behavior is to "merge and clobber" each privilege. That is, dbt will replace the less-specific set of grantees with the more-specific set of grantees.

Reviewer 2: @jtcohen6 I changed this slightly. Still work? 
```suggestion
When you set `grants` for the same model in multiple places, such as in `dbt_project.yml` and in a more-specific `.sql` or `.yml` file, dbt's default behavior replaces the less-specific set of grantees with the more-specific set of grantees.  This "merge and clobber" behavior updates each privilege when dbt builds the models. 
```
```suggestion
If `grants` are set for the same model in multiple places, such as in `dbt_project.yml` and in a more-specific `.sql` or `.yml` file, the default behavior is to "merge and clobber" each privilege. That is, dbt will replace the less-specific set of grantees with the more-specific set of grantees.
```

Author: Definitely better! I'm just going to change the last sentence very slightly

For example:

<File name='dbt_project.yml'>

```yml
models:
  +grants:
    select: ['user_a', 'user_b']
```

</File>

<File name='models/specific_model.sql'>

```sql
{{ config(grants = {'select': ['user_c']}) }}
```

</File>

As a result of this configuration, `specific_model` will grant the `select` privilege to `user_c` ONLY.

Let's say we wanted to _add_ `user_c` to the existing list of grantees receiving the `select` privilege on `specific_model`, rather than _replacing_ that list entirely. To accomplish that, we can use the `+` ("addition") symbol, prefixing the name of the privilege:

<File name='models/specific_model.sql'>

```sql

> {{ config(grants = {'+select': ['user_c']}) }}

Reviewer: so if there's a single grant config given at the top-level of the project, every other grant statement in the project needs to have the `+` prefix, else it will remove the top-level grant.

Author: Correct. Thinking here is, better safe than sorry, really — accidentally too-few grants is preferable to accidentally too many. It's an opt-in to say, "All of the existing ones, and one more too!"

```

</File>

Now, the model will grant select to `user_a`, `user_b`, AND `user_c`!

**Notes:**
- This will only take effect for privileges which include the `+` prefix. Each privilege controls that behavior separately. If we were granting other privileges, in addition to `select`, and those privilege names lacked the `+` prefix, they would continue to "clobber" rather than "add" new grantees.

> - This use of `+`, controlling clobber vs. add merge behavior, is distinct from the use of `+` in `dbt_project.yml` (shown in the example above) for defining configs with dictionary values. For more information, see [the plus prefix](https://docs.getdbt.com/reference/resource-configs/plus-prefix).

Reviewer: 🤢

Author: I know :(

the other `+` prefix is the worse one IMO. I wanted to (and still want to) replace it with `config:`, for consistency across the board: https://github.com/dbt-labs/dbt-core/issues/3663

> - `grants` is the first config to support a `+` prefix for controlling config merge behavior. Currently, it's the only one. If it proves useful, we may extend this capability to new and existing configs in the future.

Reviewer: for what other configs would this be useful? indexes?

Author: All configs today are merge-clobbered, _except_:
- `tags`
- `pre-hook`
- `post-hook`

Those are all additive-only, and there's actually no way to _remove_ or _opt out of_ tags/hooks that are inherited from `dbt_project.yml`. (The good news for hooks, at least, is that they can be defined as macros which include conditional checks)
